Source/monster.cpp

```diff
@@ -2846,29 +2843,34 @@ bool M_DumbWalk(int i, Direction md)
 
 bool M_RoundWalk(int i, Direction md, int *dir)
 {
-	if (*dir)
-		md = left[left[md]];
+	if (*dir != 0)
+		md = left[left[md]]; // 90 degrees left
 	else
-		md = right[right[md]];
+		md = right[right[md]]; // 90 degrees right
+
+	Direction turn90deg = md;
 
-	Direction mdtemp = md;
 	bool ok = DirOK(i, md);
-	if (!ok) {
-		if (*dir) {
-			md = right[mdtemp];
-			ok = DirOK(i, md) || (md = right[right[mdtemp]], DirOK(i, md));
+
+	for (int j = 0; !ok && j < 2; j++) {
+		// Keep trying up to an including backwards
+		if (*dir != 0) {
+			md = right[md];
 		} else {
-			md = left[mdtemp];
-			ok = (DirOK(i, md) || (md = left[left[mdtemp]], DirOK(i, md)));
+			md = left[md];
 		}
+		ok = DirOK(i, md);
 	}
-	if (ok) {
-		M_WalkDir(i, md);
-	} else {
-		*dir = !*dir;
-		ok = M_CallWalk(i, opposite[mdtemp]);
+
+	if (!ok) {
+		// Try 90 degrees in the oposit then desired direction

```

Comment:
Just grammar and spelling here.

```suggestion
		// Try 90 degrees in the opposite than desired direction
```